[1단계 - 자동차 경주 구현] 소주캉(강승주) 미션 제출합니다.

sojukang opened the pull request:
안녕하세요, 부족하지만 `범고래`와 함께 최선을 다해서 작성해본 미션을 리뷰어님께 리뷰요청 드립니다.
아래는 저희 페어가 구현 중 들었던 궁금증을 정리한 사항들인데, 양이 많아서 답을 주시지 않으셔도 괜찮습니다!!

1. 입출력 메시지 등을 관련 객체에 둘지 메시지를 담은 객체를 따로 만들어야 할지 궁금합니다.

2. 파라미터로 받는 객체에 `final` 붙이면 좋을까요? ex) `splitCarNames(final String input)`

3. 예외 입력시 입력 받는 부분을 다시 실행하게 했는데, 이를 `try - catch` 문에서 재귀 호출을 부르는 방식으로 구현하는 방법이 좋은 방법인지 궁금합니다.
```java
public static int getIterationNo() {
	System.out.println(GET_ITERATION_NO_MESSAGE);
	String input = scanner.nextLine();
	try {
		validIterationNo(input);
		return Integer.parseInt(input);
	} catch (RuntimeException e) {
		System.out.println(e.getMessage());
		return getIterationNo();
	}
}
```
4. `model -> controller -> view` 계층간 데이터 이동을 위해 `CarDto`를 사용하였는데, `Car`객체를 그냥 사용할 때보다 더 나은 설계가 되었는지 궁금합니다.

5. 주어진 형식으로 출력하기 위해 `View` 에서 `CarDto`를 받아서 `getter`로 렌더링하는 것과 `CarDto`에서 `toString`을 정의하여 `View`에서 이용하는 것 중 어떤 게 더 좋은 설계일지 궁금합니다.

6. 자동차 이름 검증이 `View -> Controller-> Model(Car)` 에서 이루어지는데, 좋은 설계일지 궁금합니다.

7. `CarRepository` 에서 Random 값을 생성해서 `Car`에 넘겨주는 방식에서 `CarController`가 `CarRepository`에 Random 값이 담긴 `List`를 넘겨주는 방식으로 변경하였습니다. 랜덤을 이용하는 부분을 로직에서 분리해서 테스트하기 쉽게 하려는 의도로 리팩토링을 하였는데, 개선이 되었는지 궁금합니다. 

8. 처음 구현 당시, `CarRepository` 를 `싱글톤 패턴`으로 구현하였다가 제거했습니다. `CarController` 객체를 생성하여 사용했을 때, `CarRepository`도 새롭게 생성되어야 하지 않나? 생각해서 제거를 했습니다. 혹시 자동차 도메인 미션 같은 경우, `싱글톤 패턴`을 사용하면 좋은 상황이 있을까요?

sojukang committed:
feat: JUnit 테스트 추가

sojukang committed:
docs: README 작성

sojukang committed:
feat: 기본 구분자 기능 추가

sojukang committed:
feat: 커스텀 구분자 기능 추가

sojukang committed:
feat: 빈문자 or null 판정 기능 추가

sojukang committed:
feat: 기본 구분자 합 기능 추가

sojukang committed:
feat: 커스텀 구분자 합 기능 추가

sojukang committed:
feat: 숫자 이외의 값 or 음수 예외처리 기능 추가

sojukang committed:
fix: substring 범위 오류 수정

sojukang committed:
refactor: ParameterizedTest 도입

sojukang committed:
refactor(StringCalculator): 메소드 분리

sojukang committed:
test(StringCalculator): 기본 구분자, SplitAndSum 추가

sojukang committed:
refactor(StringCalculator): 메소드 이름, 순서 변경

sojukang committed:
docs(racingcar): 기능 요구사항 작성

sojukang committed:
feat: Car 객체 이름, 상태 구현

sojukang committed:
feat: Car 객체 이름 5자 초과 or 빈 문자열 검증 기능 추가

sojukang committed:
feat: Car 이동 기능 추가

sojukang committed:
feat: 0~9 사이 random 값 반환하는 RandomNo 객체 구현

sojukang committed:
feat: N대 자동차 쉼표 구분, 공백 제거 및 생성 기능 추가

sojukang committed:
docs: 기능 구현 목록 완료 체크

sojukang committed:
feat: N대의 자동차 게임 진행 기능 추가

sojukang committed:
feat: 진행 횟수만큼 게임 진행 기능 추가

sojukang committed:
fix(README): 오타 수정 (회수 -> 횟수)

sojukang committed:
feat: 우승자 목록 생성 기능 추가

sojukang committed:
feat: 자동차 이름 유저 입력 기능 추가

sojukang committed:
feat: 진행 횟수 입력 기능 추가

sojukang committed:
feat: 각 게임 결과 출력 기능 추가

sojukang committed:
feat: 최종 우승자 출력 기능 추가

sojukang committed:
feat: 진행 횟수 예외 처리 기능 추가

sojukang committed:
feat: Car 이름 중복 예외 처리 기능 추가

sojukang committed:
feat: 시도 횟수 0 입력 예외처리 기능 추가

sojukang committed:
fix: Car 이름 예외시 CarRepository 초기화 기능 추가

sojukang committed:
refactor: 하드코딩 제거 및 final 추가

sojukang committed:
refactor: Car 객체 getter 접근 수정

sojukang committed:
refactor: CarDto 추가 및 Car getter 제거

sojukang committed:
docs: README 객체 설계 수정

sojukang committed:
refactor: carsToCarDtos 메소드 분리

sojukang committed:
refactor(InputView): 메소드 명 변경

sojukang committed:
refactor(CarRepository): size 메소드 제거

sojukang committed:
refactor: 멤버 변수에 final 추가

sojukang committed:
refactor(StringCalculator): 하드코딩 제거 및 메소드 수정

validNullOrBlank 메소드 내 isEmpty -> isBlank
isCustomInput 메소드 내 길이 판정 조건 제거
isSpecialDelimiter 메소드 로직을 각각 equals -> contains로 변환

sojukang committed:
refactor(StringCalculator): Splitter 클래스 분리

sojukang committed:
refactor: carDto 출력 toString 구현

sojukang committed:
refactor: 메소드 분리 및 로직 수정

RacingController runInIteration 메소드 분리
InputView 양수 검증 정규식 수정 및 validNonZeroInput 삭제
CarRepositoryTest afterEach에서 객체를 다시 참조하지 않도록 수정

sojukang committed:
refactor: 메소드 순서 변경 및 하드코딩 제거

sojukang committed:
refactor: 랜덤 관련 로직 수정

CarRepository 에서 Random 값을 생성해서 Car에 넘겨주는 방식에서 CarController가 CarRepository에 Random 값이 담긴 List를 넘겨주는 방식으로 변경

sojukang committed:
refactor: CarRepository singleton 패턴 제거

sojukang committed:
refactor: 필드 초기화 -> 생성자 초기화 변경, 메소드 순서 변경

sojukang committed:
test: 자동차 이름 중복 체크 테스트 추가

sojukang committed:
docs: 객체 설계 수정

sojukang committed:
refactor: Random -> ThreadLocalRandom 변경

sojukang committed:
refactor: for -> forEach

sojukang committed:
refactor: typo winners -> cars

choihz commented on src/test/java/junit/CollectionTest.java:
```diff
@@ -0,0 +1,42 @@
+
+	@Test
+	void size() {
+		assertThat(numbers.size()).isEqualTo(4);
```
테스트가 실패하네요 😅

sojukang replied:
수정하였습니다.

choihz commented on src/test/java/racingcar/controller/CarControllerTest.java:
```diff
@@ -0,0 +1,41 @@
+import org.junit.jupiter.api.Test;
+
+import racingcar.model.Car;
+import racingcar.model.CarRepository;
```
사용하지 않는 `import` 문 제거는 기본 중의 기본입니다~ 커밋 전에 반드시 확인해 주세요! 커밋을 IntelliJ에서 하신다면 자동 수행하도록 설정할 수도 있으니 방법을 찾아 보세요.

sojukang replied:
패키지 단위로 코딩 컨벤션 적용을 알아냈습니다. 😅

choihz commented on src/test/java/junit/StringTest.java:
```diff
@@ -0,0 +1,42 @@
+* Set의 contains() 메소드를 활용해 1, 2, 3의 값이 존재하는지를 확인하는 학습테스트를 구현하려한다.
+구현하고 보니 다음과 같이 중복 코드가 계속해서 발생한다.
+JUnit의 ParameterizedTest를 활용해 중복 코드를 제거해 본다.
+* */
```
여기 있을 내용이 아닌 것 같아요~

sojukang replied:
수정하였습니다.

choihz commented on src/test/java/racingcar/model/CarRepositoryTest.java:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	@Test
+	void 저장() {
```
테스트의 이름들이 테스트가 무슨 동작을 검증하는지 충분히 드러내고 있나요? 또한, `StringTest`에서 사용해본 `@DisplayName`을 사용하는 건 어떨까요? 어떤 장점이 있을까요?

sojukang replied:
@DisplayName 사용시 IDE 내 한글 메소드명으로 인한 오류가 사라지고 테스트에 관한 설명을 더 자세하게 적을 수 있다는 장점이 있어서 전체 변경하였습니다.

choihz commented on src/main/java/racingcar/model/RandomNo.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.concurrent.ThreadLocalRandom;
+
+public class RandomNo {
```
이름은 축약하지 않는 편이 좋을 것 같네요. 또한, 축약하지 않은 이름이라 해도 현재 클래스의 이름이 역할에 딱 들어맞지는 않는 것 같이 느껴져요.

sojukang replied:
RandomNo -> Random으로 바꾸고 Random.createNumbers 같이 외부에서 어떤 일을 하는지 한번에 알 수 있게 수정해보았습니다.

choihz commented on src/main/java/racingcar/model/RandomNo.java:
```diff
@@ -0,0 +1,16 @@
+
+public class RandomNo {
+	private static final int RANDOM_BOUND = 10;
+	private final ThreadLocalRandom random;
```
상수와 멤버 변수는 의미하는 바가 많이 다르기 때문에, 사이에 한 줄을 띄워 주면 눈에 더 잘 들어온답니다.

sojukang replied:
전체 수정하였습니다.

choihz commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,46 @@
+		}
+	}
+
+	private static void validIterationNo(String input) {
```
valid라는 단어는 형용사라서, 메서드의 **동작**을 제대로 드러내지 못하고 있는 것 같아요.

sojukang replied:
어떤 개발자님이 축약해 쓰시는 것을 보고 따라해보았는데, 명확하게 validate으로 바꾸는 게 좋을 것 같아 전체 수정하였습니다.

choihz commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Scanner;
+
+public class InputView {
+	public static final String POSITIVE_REGEX = "^[1-9]+[0-9]*$";
```
`public`일 필요가 있나요?

sojukang replied:
이런.. IDE 기능 이용하고 주의해서 확인하겠습니다.

choihz commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,46 @@
+	private static final String GET_CAR_NAMES_MESSAGE = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+	private static final String GET_ITERATION_NO_MESSAGE = "시도할 회수는 몇회인가요?";
+	private static final String ERROR_GET_POSITIVE_MESSAGE = "양수를 입력해주세요.";
+	private static final String ERROR_GET_NOT_EMPTY_MESSAGE = "빈 값을 입력하셨습니다. 양수를 입력해주세요.";
```
이 문자열들을 상수로 분리했을 때의 장점이 뭐라고 느끼셨는지 궁금하네요 👀

sojukang replied:
메시지 관리, 수정을 편하게 하고 하드 코딩을 줄이려는 의도로 상수로 분리하였는데, 그냥 상수화 없이 메소드에 두는 게 좋았을까요?

choihz replied:
이건 사실 리뷰어마다 다르게 생각할 거고, 취향의 문제로 볼 수도 있는데요. 저는 여러 번 사용되는 게 아니고 메시지 자체만 보고도 의미를 파악할 수 있다면 굳이 상수로 빼지 않아도 된다고 생각하는 편이에요. 상수 이름을 짓는 것도 어렵고, 또 메서드를 보면서 상수의 값을 직접 파악하려면 시선이 이동해야 하니까 탐색 비용이 든다고 볼 수도 있죠. 꼭 바꾸라는 의미는 아니고, 상수화에 지나치게 집작할 필요는 없다는 의미로 받아들이시면 좋을 것 같아요.

choihz commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,46 @@
+	private static final String GET_ITERATION_NO_MESSAGE = "시도할 회수는 몇회인가요?";
+	private static final String ERROR_GET_POSITIVE_MESSAGE = "양수를 입력해주세요.";
+	private static final String ERROR_GET_NOT_EMPTY_MESSAGE = "빈 값을 입력하셨습니다. 양수를 입력해주세요.";
+	private static final Scanner scanner = new Scanner(System.in);
```
위에 있는 상수들과는 네이밍 컨벤션이 다르군요. 선택의 기준이 어떻게 되나요?

sojukang replied:
에러 발생시 메시지와 정상 작동시 메시지를 나누려고 앞에 ERROR를 붙였는데, 에러 발생시 수행을 기준이 아니라 에러 상황에 관한 이름으로 수정하였습니다.

choihz replied:
이건 `scanner`에 관한 코멘트였는데 헷갈리게 작성되었군요 😅

sojukang replied:
`scanner`는 일반적으로 많이 사용하는 네이밍대로 아래에서도 파악하기 쉽게 `scanner`라고 이름 붙였습니다.

choihz commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	private void validDuplicationCarNames(String[] carNames) {
+		Set<String> temp = new HashSet<>(Arrays.asList(carNames));
```
`temp` 같은 변수명은 좋지 않네요 😢 맥락을 충분히 담는 이름을 고민해 보아요.

sojukang replied:
temp -> duplicationCheck으로 수정해보았습니다.

choihz commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,66 @@
+import racingcar.model.CarRepository;
+import racingcar.model.RandomNo;
+
+public class CarController {
```
소주캉은 컨트롤러의 역할이 뭐라고 생각하시나요? 현재 컨트롤러가 어떤 일들을 하고 있는지 쭉 살펴볼까요?

sojukang replied:
Controller의 역할이(토비의 스프링 참고)
1. 사용자 요청 해석(메소드가 필요로 하는 파라미터 타입에 맞게 요청 정보 수정)
2. 비즈니스 로직 위임
3. 모델 생성
4. 어떤 뷰 사용할지 결정 

이렇게 생각하였는데, CarController가 너무 많은 비즈니스 로직에 관여하던 것을 Random과 Cars에 위임하여 사용자 요청 해석과 모델 생성에 집중하도록 수정하였습니다.

choihz commented on src/main/java/racingcar/model/CarRepository.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class CarRepository {
```
이런 이름을 사용하신 이유가 궁금합니다 👀

sojukang replied:
초기 설계에서 일종의 DB를 담당하도록 생각하여  View - Controller - Service - Repository - Model 구조로 Service에서 로직을 담당하도록 구현 하려고 했습니다. 현재 어플리케이션에서는 View - Controller - Model로 충분한 것 같아 Controller와 Model에 Service의 역할을 나누고(잘못 생각) Cars와 CarRepository 정도의 이름을 고민하다가 남겨두었는데, 바꾸는게 좋을 것 같아 Cars로 수정하였습니다.

choihz commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,52 @@
+
+	public CarDto toCarDto() {
+		return new CarDto(this.name, this.position);
+	}
```
DTO를 사용하신 이유는 무엇일까요?

sojukang replied:
Domain의 Car 객체가 데이터, 검증, 이동 등의 기능을 담당하도록 자신과 관련된 책임을 가진 능동적인 객체로 만들었습니다. 불변객체가 아닌 이러한 Car 객체가 Domain을 넘어 Controller, View 로 전달되는 것은 위험하다고 생각하여 DTO를 만들어 계층간 전달하려는 의도로 만들었습니다. 페어의 리뷰어이신 `희봉`이 남겨주신 피드백을 보고 
```
Car 가 변경되면 CarDto도 변경
CarDto가 변경되면 Car는 변경되지 않음
범고래가 의도한 동작 방향일까요? CarDto는 단순히 Car의 내용을 View에서 사용하도록 전달하는 폼일 뿐인데 CarDto가 변경된다고 Car가 변경되는게 괜찮을까요? CarDto를 만든 이유에 어긋나지 않나요?
```
Car -> CarDto 의존을 CarDto -> Car 의존으로 수정하였습니다. 이때 Car에 없었던 getter 가 추가되어 '이러면 그냥 Car를 전달하는게 더 깔끔하고 좋지 않은가?' 를 생각해보았습니다. 그래도 자신의 로직을 담당하는 Car 객체를 전달하는 것보다 안전하다고 생각하여 CarDto를 사용하였습니다.

choihz commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,52 @@
+	private static final int INITIAL_POSITION = 0;
+	private static final int MOVE_CONDITION_NO = 4;
+	private static final String CAR_NAME_LENGTH_OVER_ERROR_MESSAGE = "자동차 이름이 5글자를 초과했습니다. 5글자 이하로 입력해주세요.";
+	private static final String CAR_NAME_IS_EMPTY_ERROR_MESSAGE = "자동차 이름에 빈 문자열이 입력됐습니다.";
```
> 입출력 메시지 등을 관련 객체에 둘지 메시지를 담은 객체를 따로 만들어야 할지 궁금합니다.

각 방법의 장단점을 직접 정리해 주시면 좋을 것 같아요~

sojukang replied:
관련된 메시지를 한번에 관리할 수 있다는 장점은 있지만 메시지가 사용 맥락에서 분리되고 객체가 아닌 저장소의 역할을 하게 되는 단점이 있다고 생각하였습니다.

choihz replied:
네 좋네요 👍
파악하신 장단점을 바탕으로 각 상황에 맞게 직접 판단해 주시면 됩니다.

choihz commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,66 @@
+		return Stream.of(input.split(CAR_NAME_DELIMITER))
+			.map(String::trim)
+			.toArray(String[]::new);
+	}
```
> 파라미터로 받는 객체에 `final` 붙이면 좋을까요? ex) `splitCarNames(final String input)`

붙였을 때와 안 붙였을 때의 차이가 무엇인가요? 🤨

sojukang replied:
우테코 강의 도중 한 코치님이 모든 파라미터에 final을 사용하시는 것을 보고 찾아 보았는데, 
`Java always makes a copy of parameters before sending them to methods. 
This means the final doesn't mean any difference for the calling code. 
This only means that inside the method the variables can not be reassigned.
(stackOverFlow)`
현재 사용중인 메소드 내에서의 재할당을 방지할 수 있다는 효과가 있다고 이해하였습니다. 그러나 자바를 더욱 장황하게 보이게 만들어 가독성을 해치고, 메소드가 충분히 짧게 작성되었을 경우 얻는 이익이 크지 않다고 생각하였습니다. 자바 기본 라이브러리나 스프링 내부 등 다른 코드를 살펴보고 쓰이지 않는 경우가 많은 것 같아 쓰지 않았습니다.

choihz replied:
> 현재 사용중인 메소드 내에서의 재할당을 방지할 수 있다는 효과가 있다고 이해하였습니다. 그러나 자바를 더욱 장황하게 보이게 만들어 가독성을 해치고, 메소드가 충분히 짧게 작성되었을 경우 얻는 이익이 크지 않다고 생각하였습니다. 자바 기본 라이브러리나 스프링 내부 등 다른 코드를 살펴보고 쓰이지 않는 경우가 많은 것 같아 쓰지 않았습니다.

역시 좋네요 👍

choihz commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,46 @@
+			System.out.println(e.getMessage());
+			return getIterationNo();
+		}
+	}
```
> 3. 예외 입력시 입력 받는 부분을 다시 실행하게 했는데, 이를 `try - catch` 문에서 재귀 호출을 부르는 방식으로 구현하는 방법이 좋은 방법인지 궁금합니다.

> 4. `model -> controller -> view` 계층간 데이터 이동을 위해 `CarDto`를 사용하였는데, `Car`객체를 그냥 사용할 때보다 더 나은 설계가 되었는지 궁금합니다.

> 5. 주어진 형식으로 출력하기 위해 `View` 에서 `CarDto`를 받아서 `getter`로 렌더링하는 것과 `CarDto`에서 `toString`을 정의하여 `View`에서 이용하는 것 중 어떤 게 더 좋은 설계일지 궁금합니다.

> 6. 자동차 이름 검증이 `View -> Controller-> Model(Car)` 에서 이루어지는데, 좋은 설계일지 궁금합니다.

모두 좋은 방법인지, 또는 좋은 설계인지 묻고 있는 질문이네요. 제가 답하기 전에, 소주캉이 생각하는 "좋은" 방법 또는 설계가 무엇인지를 먼저 물어보고 싶어요.

sojukang replied:
3. while보다 재귀를 사용하는 것이 더 명확하고 깔끔하여 사용하였는데, 다소 지저분한 try - catch 없이 예외 처리 - 재입력을 구현하는 방법을 페어와 생각해내기 힘들었습니다. `희봉` 님이 남겨주셨는데, 더 공부가 필요하다고 생각했습니다.
```
재입력을 받아야하는 상황을 위해 try-catch로 구현하신 것 잘하셨습니다. 👍
구현하신 것처럼 사용해도 괜찮습니다.
```
4. 위의 CarDto에서 설명 드린 이유로 도입하였고, 그 이유로 제 생각에는 더 낫다고 생각했는데, 개발 경험이 거의 없어 확신할 수 없습니다. 
5.  ```MVC 패턴의 장점은 Data를 담은 Model과 표현하는 View가 분리되어 같은 Model이라도 다른 View에 전달되면 다른 방식으로 Data가 출력될 수 있게 할 수 있다는 점이다(토비의 스프링).``` 을 보고  다시 생각해보았는데, Model과 View가 가진 역할이 명확히 분리되어 확장성 있는 설계가 좋은 설계라고 생각하여 다시 ResultView에서 getter를 통해 Data를 받아 렌더링하도록 하였습니다. 
6. 핵심 기능과 관련된 데이터라면, View에서도 1차 검증이 이루어져야하고, service logic에서도 2차 검증이 이루어져야한다고 생각합니다. 그러나 현재 어플리케이션에서는 한 번으로 족하다고 판단하였고, 그 책임을 Car와 관련된 것이라면 가장 전문가인 Car에 부여하였고, 시도 횟수 같이 전체적인 경우라면 InputView에서 검증하도록 하였습니다.  View -> Controller -> Model 로 갈수록 예외 상황을 빨리 캐치할 수 없게 되지만, 검증 책임을 관련 역할을 가진 Domain 객체에 부여하여 얻는 관리상의 이점이 있으면 도메인 객체가 가지도록 하는 것이 좋다고 생각하였습니다.

choihz replied:
사실 질문의 의도는 3~6번 각각에 대한 좋은 설계를 묻는 게 아니라, 보편적으로 좋은 설계가 무엇이라 생각하느냐, (질문자는) 어떤 것에 기준을 두고 이를 평가하느냐 묻는 거였어요. 테스트하기 좋은 코드, 유지보수하기 좋은 코드, 읽기 쉬운 코드, 성능이 좋은 코드… 뭐 많죠. 이것에 대해 서로 합의가 되지 않은 상황에서 특정 구현이 좋은 설계냐 아니냐를 논할 수 없기 때문에 남긴 질문이었어요.

리뷰어 입장에서는 '이렇게 구현했는데 좋은 방법일까요? 괜찮을까요?'라고 질문하시면 답변하기가 조금 어려워요. 어떤 관점에서 좋은 방법이냐고 묻는 것인지도 알 수 없고, 대조군 같은 것도 없다면 말이지요. 그렇기 때문에 다음에 질문하실 때는, 지금 소주캉이 다시 코멘트를 주신 것처럼 본인의 생각과 고민을 충분히 드러낸 뒤 구체적인 질문을 주신다면 조금 더 만족할 만한 답변을 얻으실 수 있을 것 같아요.

이번 질문의 경우에는 다시 주신 코멘트에서 소주캉이 드러내신 의견에 기본적으로 동의하구요, 또 다른 질문이 있으시다면 다음 단계에서 다시 주세요!

choihz commented on src/main/java/racingcar/model/CarRepository.java:
```diff
@@ -0,0 +1,56 @@
+	public void move(List<Integer> randoms) {
+		IntStream.range(0, randoms.size())
+			.forEach(i -> cars.get(i).move(randoms.get(i)));
+	}
```
> `CarRepository` 에서 Random 값을 생성해서 `Car`에 넘겨주는 방식에서 `CarController`가 `CarRepository`에 Random 값이 담긴 `List`를 넘겨주는 방식으로 변경하였습니다. 랜덤을 이용하는 부분을 로직에서 분리해서 테스트하기 쉽게 하려는 의도로 리팩토링을 하였는데, 개선이 되었는지 궁금합니다.

개선이 되었는지 여부도 본인이 판단할 수 있지 않을까요? 어떤가요? 분리를 통해 테스트를 쉽게 하려는 의도를 달성하는 데 성공하셨나요?

sojukang replied:
랜덤 값을 그대로 전달하는 방법에서 리스트를 만들어 전달하여 약간의 계산 자원을 더 이용하는 단점이 있지만 테스트가 가능하게 하여 얻는 이점이 더 많다고 생각하였습니다.

choihz commented on src/main/java/racingcar/model/CarRepository.java:
```diff
@@ -0,0 +1,56 @@
+
+	public CarRepository() {
+		this.cars = new ArrayList<>();
+	}
```
> 처음 구현 당시, `CarRepository` 를 `싱글톤 패턴`으로 구현하였다가 제거했습니다. `CarController` 객체를 생성하여 사용했을 때, `CarRepository`도 새롭게 생성되어야 하지 않나? 생각해서 제거를 했습니다. 혹시 자동차 도메인 미션 같은 경우, `싱글톤 패턴`을 사용하면 좋은 상황이 있을까요?

싱글턴 패턴에 관해 학습하셨다면 그 장점과 단점, 사용하면 좋은 상황에 대해서도 고민해 보셨을 것 같은데요. 우선 처음 구현 당시 `CarRepository`를 싱글턴 패턴으로 구현하기로 결정했던 이유가 궁금하네요.

sojukang replied:
초기 Repository를 일종의 Car 를 담은 DB와 같이 활용하기 위해 도입하였습니다. (설계상 유일, 타 클래스의 데이터 접근 용이) 그러나 설계 중 CarRepository를 능동적 객체로 이용하도록 변경하며 리팩토링하며 싱글톤이 아니게 수정하였습니다. 이 외에도 싱글톤 패턴이 프레임 워크의 도움 없이는 부작용이 많아 쓰지 말아야하겠다고 결론내었습니다.

choihz commented on src/test/java/racingcar/model/CarRepositoryTest.java:
```diff
@@ -0,0 +1,69 @@
+import org.junit.jupiter.api.Test;
+
+public class CarRepositoryTest {
+	CarRepository carRepository;
```
테스트 코드에서도 접근 제어자에 신경 써 주세요.

sojukang replied:
테스트 코드의 접근 제어자를 모두 default로 변경하였습니다.

choihz commented on src/test/java/racingcar/model/CarRepositoryTest.java:
```diff
@@ -0,0 +1,69 @@
+	void 우승자() {
+		List<Integer> numbers = new ArrayList<>(Arrays.asList(3, 4));
+		Car car1 = new Car("범고래");
+		Car car2 = new Car("소주캉");
```
중복을 제거할 수 있지 않을까요?

sojukang replied:
수정하였습니다.

choihz reviewed (changes requested):
안녕하세요, 소주캉! 첫 미션을 함께할 리뷰어 코니입니다. 👋

자동차 경주 미션을 잘 구현해 주셨네요! 🚙 또한 질문을 많이 남겨 주셨는데, 제가 무작정 답하기에는 어려운 경우가 많아 다시 질문을 던지느라, 오히려 제가 남긴 질문이 더 많아져 버렸어요 🤯 하지만 의견을 교환하는 과정에서 더 많은 걸 얻어가실 수 있으리라 생각합니다. 이외에도 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

sojukang committed:
fix(CollectionTest): 테스트 검증 값 오류 수정

sojukang committed:
fix: 부적절한 import, 주석 제거

sojukang committed:
refactor: CarController의 책임을 Cars로 이동

rename: CarRepository -> Cars, validDuplicationCarNames -> checkDuplicationCarNames
feat: Car 이름 중복 체크 이동
test: 중복 체크 이동 및 setUp으로 중복 제거

sojukang committed:
style: code convention

sojukang committed:
refactor: 테스트 접근자 public -> default, 메소드 이름 변경

한글 -> 영문으로 변경, @DisplayName 사용

sojukang committed:
refactor: 접근 제한자 범위 좁힘

sojukang committed:
refactor: RandomNo -> Random, 랜덤 리스트 생성 책임 이동

sojukang committed:
refactor: 생성자에서 createCars 호출 제거

Client인 ApplicationMain에서 RacingController에 기대하는 책임을 명확하게 하기위해 public으로 변경 후 ApplicationMain에서 호출

sojukang committed:
refactor: 축약어 수정, validateIterationNumber 접근자 변경

No -> Number
valid -> validate

sojukang committed:
refactor: 상수와 멤버 변수 사이 개행

sojukang committed:
refactor: carDto 렌더링 책임 toString -> ResultView

sojukang committed:
refactor: 메소드명, 상수명 및 메소드 순서 변경

sojukang committed:
refactor: 테스트 이름을 명확하게 변경

sojukang committed:
refactor: Car -> CarDto 의존을 CarDto -> Car로 변경

sojukang committed:
refactor: checkSet -> duplicationCheck

sojukang commented:
상세히 체크하고 리뷰 해주셔서 정말 감사드립니다. 남겨주신 질문들을 생각해보며 다양한 자료들을 참고하고 공부하는 시간이 되어 유익하였습니다. 비전공자로 프리코스에서 자바를 처음 접하고 2달 남짓 시간 동안 공부해보았는데, 아직은 어떤 설계가 더 좋은 설계인지 감이 잡히지 않고 막막하지만 이번 계기로 많이 생각해볼 수 있는 시간을 가졌습니다.

choihz commented on src/main/java/racingcar/ApplicationMain.java:
```diff
@@ -0,0 +1,11 @@
+public class ApplicationMain {
+	public static void main(String[] args) {
+		RacingController racingController = new RacingController();
+		racingController.createCarsByUserInput();
+		racingController.run();
```
사용자 입력을 받아야 하는 부분이 1. 자동차 이름 2. 게임 횟수 이렇게 둘 있죠. 그런데 자동차 이름을 입력받기 위해서는 명시적으로 `createCarsByUserInput()` 메서드를 실행해야 하고 게임 횟수를 받는 부분은 `run()`에 숨어 있네요. 이렇게 구현하신 이유가 있을까요?

choihz commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,44 @@
+		this.random = new Random();
+	}
+
+	String[] splitCarNames(String input) {
+		return Stream.of(input.split(CAR_NAME_DELIMITER))
+			.map(String::trim)
+			.toArray(String[]::new);
+	}
```
이 메서드를 비롯하여 전반적으로, 과연 컨트롤러의 역할인가 의문이 드는 부분이 있는 것 같아요. 또한 `RacingController`와 `CarController`가 나누어진 기준도 다소 모호해 보이는데, 혹시 그 기준을 설명할 수 있을까요?

choihz commented on src/test/java/racingcar/controller/CarControllerTest.java:
```diff
@@ -0,0 +1,29 @@
+
+class CarControllerTest {
+
+	@Test
+	@DisplayName("자동차 이름 콤마 기준 분리, 공백 제거")
+	void splitCarNamesTest() {
```
이 기능을 테스트하는 코드가 과연 컨트롤러 테스트에 위치하는 게 맞을지도 함께 고민하면 이해하는 데 좀 더 도움이 될 것 같아요.

choihz commented on src/test/java/racingcar/model/CarsTest.java:
```diff
@@ -0,0 +1,64 @@
+import org.junit.jupiter.api.Test;
+
+class CarsTest {
+	Cars cars;
+	List<Integer> moveConditionNumbers;
```
`private`이어도 되지 않을까요~?

sojukang replied:
수정하였습니다.

choihz commented on src/test/java/racingcar/model/CarsTest.java:
```diff
@@ -0,0 +1,64 @@
+
+	@BeforeEach
+	void setUp() {
+		moveConditionNumbers = new ArrayList<>(Arrays.asList(3, 4));
```
`new ArrayList()`로 다시 감싸는 이유가 무엇일까요?

sojukang replied:
수정하였습니다.

choihz commented on src/test/java/racingcar/model/CarsTest.java:
```diff
@@ -0,0 +1,64 @@
+		cars.move(moveConditionNumbers);
+		List<CarDto> expected = Arrays.asList(new CarDto(carStop), new CarDto(carMove));
+
+		assertThat(cars.getCars().equals(expected)).isTrue();
```
```suggestion
		assertThat(cars.getCars()).isEqualTo(expected);
```

이 방식의 단언문이 더 이해하기 쉽지 않을까요? 😅

sojukang replied:
isEqualTo가 equals를 검증하는 것을 잠깐 잊고있었네요😅

choihz commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,57 @@
+package racingcar.model;
+
+class Car {
```
지금 메서드 배치 순서가 default → private 순인 것으로 보이는데, 이렇게 배치했을 때 어떤 장점을 느끼셨을까요?

sojukang replied:
처음에는 넓은 scope 순서로 배치하는 것이 위에 다른 클래스에서 사용하는 메소드가 위치하고, 아래 클래스 내부에서만 이용하는 메소드가 위치하여 좋다고 생각하였습니다. 코니가 답변주신 질문을 보고 일반적인 컨벤션이 있나 찾아보았는데, 

> These methods should be grouped by functionality rather than by scope or accessibility. For example, a private class method can be in between two public instance methods. The goal is to make reading and understanding the code easier.
(Code Conventions for the java(Oracle)

> The order you choose for the members and initializers of your class can have a great effect on learnability. However, there's no single correct recipe for how to do it; different classes may order their contents in different ways.
> What is important is that each class uses some logical order, which its maintainer could explain if asked. For example, new methods are not just habitually added to the end of the class, as that would yield "chronological by date added" ordering, which is not a logical ordering.
> (Google Java Style Guide)

글을 보고 scope 보다는 메소드의 기능별로 응집도 있도록 논리적으로 가독성 있게 배치하는 것이 옳은 배치라고 생각하였습니다. 2단계부터 적용해보겠습니다.

choihz commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,67 @@
+		return carsToCarDtos(winnerCars);
+	}
+
+	private List<CarDto> carsToCarDtos(List<Car> cars) {
+		return cars.stream()
+			.map(CarDto::new)
+			.collect(Collectors.toList());
+	}
```
> Car -> CarDto 의존을 CarDto -> Car 의존으로 수정하였습니다. 이때 Car에 없었던 getter 가 추가되어 '이러면 그냥 Car를 전달하는게 더 깔끔하고 좋지 않은가?' 를 생각해보았습니다. 그래도 자신의 로직을 담당하는 Car 객체를 전달하는 것보다 안전하다고 생각하여 CarDto를 사용하였습니다.

지금 코드를 보면 `Car`는 `CarDto`를 모르게 되었지만, `Cars`는 `CarDto`를 잘 알고 있는 것 같아요. 이건 의도하신 부분일까요?

choihz commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,47 @@
+
+	private static void validatePositiveInput(String input) {
+		if (!input.matches(POSITIVE_REGEX)) {
+			throw new RuntimeException(ERROR_NOT_POSITIVE_MESSAGE);
```
`RuntimeException`은 너무 상위의 예외라서, 조금 더 구체적인 예외를 사용해도 좋을 것 같아요.

sojukang replied:
현재 RuntimeException을 던지는 경우가 IllegalArgumentException에 해당하는 경우라고 판단하여 수정하였습니다.

choihz reviewed (approved):
안녕하세요, 소주캉! 리뷰어 코니입니다.

리뷰 반영하신 부분과 질문에 남겨주신 답 모두 잘 확인했어요. 이번에 리뷰드린 부분은 다음 단계에서 함께 반영해도 될 것 같아서 머지하도록 하겠습니다.

다음 단계에서 다시 만나요~ 👋
